src/dns.rs

```diff
@@ -0,0 +1,97 @@
+mod local;
+mod default;
+mod cache;
+
+use crate::types::Result;
+use std::net::IpAddr;
+use std::time;
+
+/// A single IP entry in the DNS cache. Each ip entry has an expiration time (based on the TTL)
+/// and a last-used timestamp. When the entry is expired, the entry is found to be stale and should
+/// be refreshed by the resolver. When expired == 0, the entry is considered to be a local override
+/// and should not be refreshed by the resolver.
+/// @todo: Does that information belong here? Or should it be in the cache layer?
```

Comment:
as stated in the todo: i'm not sure if both expired and last-used should be part of the ip entry. I think the only place where ttl / expiration is important, would be the caching layer. This means the resolver should return a tuple (ipaddress, ttl) so the layer above can cache it accordingly. Time last used may not be needed.